### DISCUSSION TITLE
Router Breadcrumbs (Displayable Routes)

### DISCUSSION THREAD

**Discussion by allan-vera** [TanStack/router#962] [Ideas]
Hello, 

First time suggesting a feature so please let me know if you need any more details. 

For my Router Layout Component, I am doing something like this.

```
<div>
   <Breadcrumbs breadcrumbs={breadcrumbs} />
   <Outlet />
</div>
```

And, I am currently generating breadcrumbs like so:

```
const router = useRouterState();
const { matches } = router;
const pathNames = matches.map((match) => match.routeId);
const breadcrumbs: string[] = condensePathNames(routeIds);
```

```
export function condensePathNames(pathNames: string[]) {
  const orderedPathNames = new Set<string>();

  for (let i = 0; i < pathNames.length; i++) {
    let pathName = pathNames[i];

    // React router changes "/" to "__root__", so change back
    if (pathName === "__root__") pathName = "/";

    // ignore paths that end in / unless its "/"
    if (pathName !== "/" && pathName[pathName.length - 1] === "/") break;

    // ignore paths that are already added
    // (shouldn't be nec but just in case)
    if (orderedPathNames.has(pathName)) break;

    orderedPathNames.add(pathName);
  }

  return Array.from(orderedPathNames);
}
```

And in my Breadcrumbs component I have to do something like this to replace paramIds with their actual value
```
const BREADCRUMBS_MAP: Record<string, string> = {
  // DASHBOARD
  "/": "Dashboard",

  // USERS
  "/users": "Users",
  "/users/$userId": "User Detail",
  "/users/create": "Create User",
  "/users/$userId/edit": "Edit User",
  }

<BreadcrumbLink
                  as={i === paths.length - 1 ? undefined : Link}
                  to={replacePathParamWithValue(
                    path,
                    matches[Object.keys(matches).length - 1].params ?? null
                  )}
                >
                  {BREADCRUMBS_MAP[path as string]}
                </BreadcrumbLink>
                
 export function replacePathParamWithValue(
  path: string,
  params: Record<string, string>
) {
  // check to see if path includes $
  if (!path.includes("$") || !params) return path;

  const rawPaths = path.split("/");
  const paramKeys = Object.keys(params);
  const res = [];
  for (let i = 0; i < rawPaths.length; i++) {
    let rawPath = rawPaths[i];
    if (rawPath.includes("$")) {
      paramKeys.forEach(
        (pk: string) => (rawPath = rawPath.replace(`$${pk}`, params[pk]))
      );
    }
    res.push(rawPath);
  }

  return res.join("/");
}               
```

It works decently but as you can probably imagine, it's very fragile especially when I attempt to work with static nested paths (ie. `['/', '/users/$userId', '/users/$userId/']` <-- not sure why but '/users' got skipped. Is there currently a way I could easily get breadcrumbs from the router? If not, this is my suggested feature. 

Ideally, we could be provided a way to provide displayable names for each of the routes, that way we could retrieve that displayable name of the route that could then be used for the breadcrumbs.

Thanks!

**Comment by allan-vera** [comment]
kind bump

**Comment by snewell92** [comment]
I'm imagining this as two simple utilities as an API to build breadcrumbs; a hook and a link component. Something like 

**useBreadcrumbs**

```ts
interface BreadcrumbRoute {
  path: string;
  name: string; // where does this come from?
  // TODO any more?
}

export function useBreadcrumbs(): Array<BreadcrumbRoute> {
  // TODO MAGIC reactive

  return collectedBreadcrumbs;
}
```

and

**BreadcrumbLink**

```tsx
type Props = PropsWithChildren<{link: BreadcrumbRouter}>;

export function BreadcrumbLink({ link, children }: Props) {
  // custom onclick handling that pops history

  // parent could pass name? do we just need path then
  return <Link to={link.path}>{children}</Link>;
  // can we use Link internally or since we are essentially pressing the browser's
  //  back button n times.
}
```

The "press browser back btn n times" feels correct, but probably has subtle things that are wrong about it (ie can you suspend routing or batch n back btn presses?).

**Reply by EwenQuim** [reply]
`useBreadcrumbs()` should definitely be in the library by default.

Not sure about that `BreadcrumbLink` component. Let the user code it for their needs?

**Comment by tigawanna** [comment]
This works for me 
```tsx
//use-tsr-breadcrumbs
import { useLocation } from "@tanstack/react-router";

export function useTSRBreadCrumbs() {
  const current = useLocation();

  const route_history = current.pathname
    .split("/")
    .filter((x) => x && x.length > 0);

  const breadcrumb_routes = route_history.reduce(
    (acc: { name: string; path: string }[], route) => {
      const prev_path = acc[acc.length - 1]?.path ?? "";
      acc.push({ name: route, path: `${prev_path}/${route}` });
      return acc;
    },
    [],
  );
  return { breadcrumb_routes };
}

```

```tsx
//TSRBreadCrumbs.tsx

import { Link } from "@tanstack/react-router";
import { useTSRBreadCrumbs } from "./use-bread-crumbs";
import { ChevronRight } from "lucide-react";

interface TSRBreadCrumbsProps {}

export function TSRBreadCrumbs({}: TSRBreadCrumbsProps) {
  const { breadcrumb_routes } = useTSRBreadCrumbs();
  if(breadcrumb_routes.length<2)return null
  return (
    <div className="w-full flex flex-wrap md:justify-end gap-0.1 p-1">
      {breadcrumb_routes.map((crumb) => {
        if (breadcrumb_routes.length - 1 === breadcrumb_routes?.indexOf(crumb)) {
          return (
            <span key={crumb.path} className="text-xs bread-crumbs-links">
              {crumb.name}
            </span>
          );
        }
        return (
          <span key={crumb.path} className="flex gap-1 justify-center items-center">
            <Link className="text-xs bread-crumbs-links" to={crumb.path}>
              {crumb.name}
            </Link>
            <ChevronRight className="size-4" />
          </span>
        );
      })}
    </div>
  );
}



```

**Reply by allan-vera** [reply]
Going to be revisiting this at some point next week and will be able to take a deeper look then, but does this take ids into account as well? For example, `users/:user_id/edit` broken down to "Dashboard / Users / User Detail / Edit User"

**Reply by itsmejawad** [reply]
This is perfect!

**Reply by tigawanna** [reply]
> Going to be revisiting this at some point next week and will be able to take a deeper look then, but does this take ids into account as well? For example, `users/:user_id/edit` broken down to "Dashboard / Users / User Detail / Edit User"

yup , it segments the whole path search params and params included

**Reply by allan-vera** [reply]
Sweet!! I'll look into refactoring my existing code at some point so it can adopt this. Also, it appears useLocation does not exist in the current version of tanstack router i am using so will need to upgrade that as well. 

Thanks for taking a look!

**Comment by chorobin** [comment]
We've been trying to address breadcrumbs recently with some experimental changes with `useMatches`. The aim is to provide a type safe way to build breadcrumbs and similar things

Firstly `useMatches` returns discriminated union so its possible narrow matches based on `fullPath` or `routeId`. Secondly there is a `isMatch` type predicate that allows you filter and narrow matches based on the shape of a match. `matches.filter(match => isMatch(match, 'context.crumb'))` will narrow matches that have the `crumb` property in `context`

https://github.com/TanStack/router/pull/2058

We have `Breadcrumbs` now in the kitchen sink example

https://github.com/TanStack/router/blob/main/examples/react/kitchen-sink-file-based/src/components/Breadcrumbs.tsx

It would be great to get some feedback

**Reply by snewell92** [reply]
Could TSR provide a hook for this if we could configure the key to mark 'crumb' (or should we just say TSR crumb API is `{ crumb: 'Name' }`)? This feels like a `useCrumbs` is waiting to be added! Although it's pretty small, so very easy to copy in.

**Reply by chorobin** [reply]
A `useCrumbs` hook would be implemented fairly easily with `useMatches` and `isMatch`

```tsx
const useCrumbs = () => {
  return useMatches()
    .filter((match) => isMatch(match, 'context.crumb'))
    .map((match) => match.context.crumb)
}
```

For the moment I'm on the side of being more flexible. `useCrumbs` hook is very specific and I want to be open to different use cases other than breadcrumbs. `useMatches` and `isMatch` are ways to create breadcrumbs but can be used for other things.

For example, maybe you have titles which you can get from loader data. This hook builds a titles array from matches which have titles in `loaderData`

```tsx
const useTitles = () => {
  const matches = useMatches()
  const titles = []

  for (const match of matches) {
    if (isMatch(match, 'loaderData.title') && match.loaderData?.title) {
      titles.push(match.loaderData.title)
    }
  }

  return titles
}
```

Simply put.

`useMatches` gives you a union of all possible matches

`isMatch` provides a way to narrow those matches, is composable and keeps the context of the match. Its possible to also grab other things from a narrowed `match` provided the narrowing results in a `match` which has it. For example if the narrowed match also has `params`

```tsx
const useCrumbs = () => {
  return useMatches()
    .filter((match) => isMatch(match, 'context.crumb'))
    .map((match) => `${match.context.crumb} - ${match.params.postId}`)
}
```

You can also compose `isMatch` together

```tsx
const useCrumbs = () => {
  return useMatches()
    .filter(
      (match) =>
        isMatch(match, 'context.crumb') && isMatch(match, 'params.postId'),
    )
    .map((match) => `${match.context.crumb} - ${match.params.postId}`)
}
```

**Reply by NateHawk85** [reply]
First, I love the progress, and this library has made 99% of my issues with routing disappear. Unfortunately, Breadcrumbs is a big one that I haven't found a solid way past... yet. In my eyes, the core of this Breadcrumb issue is that `useMatches` doesn't return index routes, nor does there seem to be an option to include them in the results. For example, if I have a directory structure of

```
routes/
├─ __root.tsx
|
├─ teams/
│  ├─ index.tsx
│  ├─ $teamId.tsx
│  |
│  ├─ $teamId.players/
│  │  ├─ index.tsx
│  │  ├─ $playerId.tsx
```

When I run `useMatches()` within my `Breadcrumbs` component, I get 3 matches (vs 5 expected) with route ids:
- `__root__`
- (MISSING) `/teams/`
- `/teams/$teamId`
- (MISSING) `/teams/$teamId/players/`
- `/teams/$teamId/players/$playerId`

One implementation could be an `includeIndex` option flag (or something similar) for `useMatches` that would include those index routes when evaluating matches.

**Reply by chorobin** [reply]
It sounds like you want branches, not index routes? I think you need `route.tsx` here aswell as an index route for the branch to match

**Reply by gabbelobo** [reply]
Just to clarify for others having the same problem, to get the breadcrumbs to work properly, have the following structure:
```
routes/
├─ __root.tsx
|
├─ clients/
│  ├─ index.tsx
│  ├─ route.tsx
│  ├─ create.tsx
```
index.tsx
```
import { createFileRoute } from '@tanstack/react-router'

export const Route = createFileRoute('/clients/')({
  component: () => <div>Hello from '/clients</div>,
})
```

route.tsx
```
import { createFileRoute } from '@tanstack/react-router'

export const Route = createFileRoute('/clients')({
  loader: () => ({
    crumb: 'Clients',
  }),
})
```

create.tsx
```
import { createFileRoute } from '@tanstack/react-router'

export const Route = createFileRoute('/clients/create')({
  component: () => <div>Hello from '/clients/create</div>,
  loader: () => ({
    crumb: 'Create',
  }),
})
```

And the `useMatches` will match both index and create routes properly.

**Reply by Saidmuhammad2002** [reply]
is there any way to achieve same behaviour without making route file for each page

**Reply by tigawanna** [reply]
Hook

```tsx

import { useLocation } from "@tanstack/react-router";

export function useTSRBreadCrumbs() {
  const current = useLocation();

  const route_history = current.pathname
    .split("/")
    .filter((x) => x && x.length > 0);

  const breadcrumb_routes = route_history.reduce(
    (acc: { name: string; path: string }[], route) => {
      const prev_path = acc[acc.length - 1]?.path ?? "";
      acc.push({ name: route, path: `${prev_path}/${route}` });
      return acc;
    },
    [],
  );
  return { breadcrumb_routes };
}

```


Usage

```tsx

import { Link } from "@tanstack/react-router";
import { useTSRBreadCrumbs } from "./use-tsr-breadcrumbs";

import {
  Breadcrumb,
  BreadcrumbItem,
  BreadcrumbList,
  BreadcrumbPage,
  BreadcrumbSeparator,
} from "@/components/shadcn/ui/breadcrumb";
interface TSRBreadCrumbsProps {}

export function TSRBreadCrumbs({}: TSRBreadCrumbsProps) {
  const { breadcrumb_routes } = useTSRBreadCrumbs();
  if (breadcrumb_routes.length < 2) return null;
  return (
    <div className="gap-0.1 flex w-full flex-wrap p-1 px-3 md:justify-end">
      <Breadcrumb>
        <BreadcrumbList>
          {breadcrumb_routes.map((crumb) => {
            if (
              breadcrumb_routes.length - 1 ===
              breadcrumb_routes?.indexOf(crumb)
            ) {
              return (
                <BreadcrumbItem key={crumb.path}>
                  <BreadcrumbPage className="hover:text-accent-text line-clamp-1 cursor-pointer text-xs hover:max-w-fit hover:duration-300 hover:animate-in hover:fade-in">
                    {crumb.name}
                  </BreadcrumbPage>
                </BreadcrumbItem>
              );
            }
            return (
              <div className="flex items-center gap-2" key={crumb.path}>
                <BreadcrumbItem key={crumb.path}>
                  <Link
                    to={crumb.path}
                    className="hover:text-accent-text line-clamp-1 cursor-pointer text-xs hover:max-w-fit hover:duration-300 hover:animate-in hover:fade-in"
                  >
                    {crumb.name}
                  </Link>
                </BreadcrumbItem>
                <BreadcrumbSeparator className="hidden md:block" />
              </div>
            );
          })}
        </BreadcrumbList>
      </Breadcrumb>
    </div>
  );
}


```

**Reply by protzman** [reply]
@tigawanna thanks for this. I'm curious with this setup the way it is, are you/what would be the best way to pass custom names. I have a fair number of root level routes in my app that have children that are dynamic. A file route called '/platform' for example. I don't want to just toUppercase it/everything because the child routes are explicitly case sensitive alphanumeric combinations. Seems like the older way which I had in an older version of the app was using useMatches but it doesn't seem like you're doing that. thoughts?

**Reply by IanVS** [reply]
I'm going down this same road too right now, finding that to get breadcrumbs working as shown in https://tanstack.com/router/latest/docs/guide/router-context#processing-accumulated-route-context, it's necessary to use `route.ts`  files, which makes sense.  However, the `index.ts` files _also_ have a context, so I seem to need some way to filter those out to avoid duplicate breadcrumbs.  Has anyone else faced and solved that issue?

**Reply by JClackett** [reply]
> I'm going down this same road too right now, finding that to get breadcrumbs working as shown in https://tanstack.com/router/latest/docs/guide/router-context#processing-accumulated-route-context, it's necessary to use `route.ts`  files, which makes sense.  However, the `index.ts` files _also_ have a context, so I seem to need some way to filter those out to avoid duplicate breadcrumbs.  Has anyone else faced and solved that issue?

just don't add a crumb value to the index file, the layout contains all sub crumbs then any leaf pages add their crumb values

**Reply by IanVS** [reply]
Right, so the challenge is that the leaf pages will get the context / staticData from the route as well.  So, say you have a single page, `/home`, with files like:

```
/home/route.ts // <-- Setting context.crumb here
/home/index.tsx
```

You'll get two matches, each with a context with a crumb.  So you won't be able to determine if this is a nested route or not.  I think the only way I've seen to handle this so far is to set a crumb in the index explicitly setting it to undefined.

**Reply by JClackett** [reply]
you just dont set it in index.tsx

```
/home/route.ts // crumb: "Home"
/home/index.tsx // no crumb
/home/another.tsx // crumb: "Another"
```

then wherever you are displaying your crumbs, just loop over the array of `crumbs.filter(Boolean)`

so youll get

```
Index => Home
Another => Home / Another
```

But what are you expecting/trying to display?

**Reply by IanVS** [reply]
OK, I learned `context` is merged from layout routes into leaf routes, whereas `staticData` and `loaderData` are not.  So, if you try to use context, as the docs show, you get duplicates.  Here's a dirty example: https://stackblitz.com/edit/github-atelt7ni-fd64jvxp?file=src%2FuseBreadcrumbs.ts.  Click on the "products" link, and you'll see duplicated breadcrumbs at the top.

**Reply by JClackett** [reply]
yeah I use loaderData for this reason, also because sometimes a page may fetch stuff in a loader so I can reuse in the breadcrumb label

**Comment by prc5** [comment]
I created these breadcrumbs for my app. They’re a bit rough and hacky, but hopefully, they’ll be helpful to you! 😄

```ts
  const matches = useRouterState({ select: (s) => s.matches });
  const router = useRouter();

  const indexRoute = {
    title: "Your app",
    pathname: "/",
  };

  const match = matches[matches.length - 1];

  const routes = useMemo(() => {
    const matched = router.flatRoutes
      .filter((route) => route.options.beforeLoad)
      .filter((route) => {
        if (route.fullPath === "/") {
          return match.fullPath === "/";
        }

        return match.fullPath.includes(route.fullPath);
      });

    return matched
      .filter((item, index) => {
        const prev = matched[index - 1];
        if (prev) {
          return prev.fullPath !== `${item.fullPath}/`;
        }
        return true;
      })
      .sort((a, b) => a.fullPath.length - b.fullPath.length);
  }, [match]);

  const breadcrumbs = useMemo(() => {
    const values = routes.map((item) => {
      const { id, fullPath, options } = item as AnyRoute;

      const context = options.beforeLoad?.();

      const title =
        id === "__root__"
          ? "Your app"
          : (context as { title?: string })?.title ||
            fullPath.split("/").join("") ||
            "Dashboard";

      return {
        title,
        pathname: fullPath,
      };
    });
    values.unshift(indexRoute);
    return values;
  }, [routes]);

```

**Comment by ChuckJonas** [comment]
This thread save me from 2 hour-long session trying to get breadcrumbs to work.  I was following the [docs around using the RouterContext](https://tanstack.com/router/latest/docs/framework/react/guide/router-context#processing-accumulated-route-context), which is absolutely a cursed approach.  

1. routes inherit the context so you end up with duplicate crumbs if you have "pathless routes" or multiple matches
2. you can't set the names using async data

**Comment by njoshi22** [comment]
I think just being able to set a "name" property on the Route would be huge for the library.

**Reply by smashah** [reply]
yeah a simple 
```
{
...
getTitle: (context)=>`some title ${context.id}`
...
}
```
would be fine.

**Comment by kenmlee** [comment]
A simple solution:
```
// in __root.tsx
...
export const Route = createRootRouteWithContext<{
  breadcrumb?: string
  ...
}>

// in each route/index file which you want it to be in the breadcrumb list
export const Route = createFileRoute("/users/list/")({
  component: UserListComponent,
  beforeLoad: () => {
    return { breadcrumb: 'User List' } // or any words you want to show as a breadcrumb
  }
  ...

// In Breadcrumb Component
const Breadcrumb = () => {
  const matches = useMatches()
  const breadcrumbs = matches.filter(m => m.context.breadcrumb).map(
    m=> ({ url: m.routeId, breadcrumb: m.context.breadcrumb })
  )

return (
  <div>{breadcrumbs.map(({url,breadcrumb})=> (...)}</div>
)

// If you have both route.tsx and index.tsx, you will get duplicate items
// then write a array dedup function to make it unique
const dedupArray = (arr: Array<Record<string, unknown>>, key: string) => {
  const map = new Map();
  for (const el of arr) {
    if (!map.has(el[key])) {
      map.set(el[key], el);
    }
  }
  return Array.from(map.values());
};
```

**Comment by uap-dev** [comment]
Would love to see a bump on this, just to get this working according to [docs](https://tanstack.com/router/latest/docs/framework/react/guide/router-context#processing-accumulated-route-context). 

I have the following beforeLoad

```ts
beforeLoad: async ({ location }) => {
    const data = { pathname: location.pathname }
    await requireAuth({ data })

    return {
      breadcrumb: {
        getTitle: (match: any) => {
          return (
            match.loaderData.organization.name
          )
        },
      },
    }
  },
  ```
  
When I set breakpoints in the dev server, I can see that context.breadcrumb.getTitle exists, but on the frontend when the loader runs, context.breadcrumb.getTitle, the function is undefined

**Reply by uncvrd** [reply]
I know it's late but I just wanted to mention If anyone sees this, I think they removed the ability to pass functions in beforeLoad and you can only do it in the root context. Please correct me if I'm wrong though

**Comment by raduliviu** [comment]
Here's how I managed to get it working:

```tsx
/** 
Directory tree looks like:
src
- main.tsx
- routes
  - __root.tsx
  - index.tsx
  - posts
    - $postId.tsx
    - index.tsx
    - route.tsx
*/

// in src/routes/__root.tsx:
...
export const Route = createRootRoute({
    component: RootComponent,
    loader: () => ({ crumb: "Home" }),
})
...

// in src/routes/posts/index.tsx:
...
export const Route = createFileRoute("/posts/")({
// ^^note the trailing slash, compared to the next file
    component: RouteComponent,
})
...

// in src/routes/posts/route.tsx, this is the entire contents of the file:
import { createFileRoute, Outlet } from "@tanstack/react-router"

export const Route = createFileRoute("/posts")({
    component: RouteComponent,
    loader: () => ({ crumb: "Posts" })
})

function RouteComponent() {
    return <Outlet />
}

// in src/routes/posts/$postId.tsx:
...
export const Route = createFileRoute("/posts/$postId")({
    component: RouteComponent,
    loader: ({ params }) => ({ crumb: `Post ID: ${params.postId}` }),
})
...

// my Breadcrumbs component:
export const Breadcrumbs = () => {
    const matches = useMatches()

    const matchesWithCrumbs = matches
        .filter((match) => isMatch(match, "loaderData.crumb"))
        .map((match) => ({
            to: match.pathname,
            label: match.loaderData!.crumb,
        }))

    return (
        <nav className="flex flex-row items-center gap-1">
            {matchesWithCrumbs.map((item, index) => (
                <Link to={item.to} key={`breadcrumb-${item.to}`}>
                    {/* Your Breadcrumb styling component */}
                </Link>
            ))}
        </nav>
    )
}
```
In general, I'm not a fan of having the `src/routes/posts/route.tsx` file that's basically just for the purposes of adding the crumb data. If I were to add `loader: () => ({ crumb: "Posts" }),` directly to my `src/routes/posts/index.tsx` file, my Breadcrumb UI properly shows Home > Posts while I'm on the `/posts` route but once I go to `/posts/$postId` it disappears, leaving me with Home > Post ID: 1.

@chorobin I have some feedback:
- the only thing in the official docs regarding Breadcrumbs is in the Router Context section, which provides barely any workable example
- the sample in the [Kitchen Sink File Based](https://tanstack.com/router/latest/docs/framework/react/examples/kitchen-sink-file-based?path=examples/react/kitchen-sink-file-based/src/components/Breadcrumbs.tsx) example uses a completely different pattern than the Context
- that example relies on the `route.tsx` Layout Route pattern, which could be better highlighted/documented

**Reply by dawsonbooth** [reply]
This is what we went with as well, and I share the grievance of needing an additional `route.tsx` file at every level (minus the leaf*). Some of our routes are fairly nested, so this adds a lot of boilerplate which otherwise wouldn't be necessary.

*though it's arguably easier to maintain in a large route hierarchy if the breadcrumbs are _always_ added to a layout route even if not strictly required by the framework.

However, I can't really see an alternative with file-based routing since applying to all sub-routes is the very purpose of [layout routes](https://tanstack.com/router/v1/docs/framework/react/routing/routing-concepts#layout-routes) – that is, aside from a dedicated entrypoint for "naming" routes as @njoshi22 suggested.

**Comment by JClackett** [comment]
I have an example here, it's all working super nice and im really happy with how you set the breadcrumbs. 

https://github.com/JClackett/tss-breadcrumbs

**Comment by CBell045** [comment]
Here is how I did it with meta / titles:


```
import { Link, useRouterState } from "@tanstack/react-router";
import { Fragment, type JSX } from "react";
import {
  Breadcrumb,
  BreadcrumbItem,
  BreadcrumbLink,
  BreadcrumbList,
  BreadcrumbPage,
  BreadcrumbSeparator,
} from "@/components/ui/breadcrumb";
import { Separator } from "../ui/separator";
import { SidebarTrigger } from "../ui/sidebar";

export function Header({
  children,
}: {
  children?: React.ReactNode;
}): JSX.Element {
  const matches = useRouterState({ select: (s) => s.matches });

  const breadcrumbs = matches.flatMap(({ pathname, meta }) => {
    const title = meta?.find((m) => m?.title)?.title;
    return title ? [{ title, path: pathname }] : [];
  });

  return (
    <header className="flex h-16 shrink-0 items-center gap-2 px-2 justify-between">
      <div className="flex items-center gap-2 px-4">
        <SidebarTrigger className="-ml-1" />
        <Separator
          orientation="vertical"
          className="mr-2 data-[orientation=vertical]:self-auto data-[orientation=vertical]:h-4"
        />
        <Breadcrumb>
          <BreadcrumbList>
            {breadcrumbs.map((crumb, index) => {
              const isLast = index === breadcrumbs.length - 1;

              return (
                <Fragment key={crumb.title}>
                  <BreadcrumbSeparator className="hidden md:block first:hidden" />
                  <BreadcrumbItem className="hidden md:block last:block">
                    {isLast ? (
                      <BreadcrumbPage>{crumb.title}</BreadcrumbPage>
                    ) : (
                      <BreadcrumbLink
                        render={<Link to={crumb.path}>{crumb.title}</Link>}
                      />
                    )}
                  </BreadcrumbItem>
                </Fragment>
              );
            })}
          </BreadcrumbList>
        </Breadcrumb>
      </div>
      <div className="flex items-center px-2">{children}</div>
    </header>
  );
}
```



route.tsx:
```
import { createFileRoute } from "@tanstack/react-router";

export const Route = createFileRoute("/_main/groups")({
  head: () => ({
    meta: [{ title: "Groups" }],
  }),
});
```

index.tsx:
```
export const Route = createFileRoute("/_main/groups/$groupId/")({
  loader: ({ context, params }) =>
    context.queryClient.ensureQueryData(
      orpc.group.getById.queryOptions({ input: params.groupId }),
    ),
  head: ({ loaderData }) => ({
    meta: [{ title: loaderData?.name || "Group Details" }],
  }),
  component: RouteComponent,
});

function RouteComponent() {
...
```

**Reply by mdnjohn** [reply]
Thank you very much for this; it is great for me!

**Comment by fuerst** [comment]
Thanks for all the inspiration! I wanted to share my fairly simple solution. Maybe I missed something.

It's based on [Static Route Data](https://tanstack.com/router/latest/docs/framework/react/guide/static-route-data).
Breadcrumb of the current active route will not be shown.

src/router.tsx
```typescript
declare module '@tanstack/react-router' {
	interface StaticDataRouteOption {
		breadcrumbTitle?: string;
	}
}
```

src/components/Breadcrumbs.tsx
```typescript
import { Link, useMatches } from '@tanstack/react-router';

export function Breadcrumbs() {
	const matches = useMatches();
	const currentFullPath = matches[matches.length - 1].fullPath;

	const items = matches
		.filter(
			(match) =>
				match.staticData?.breadcrumbTitle && match.fullPath !== currentFullPath,
		)
		.map((match) => (
			<li key={match.id}>
				<Link to={match.fullPath} params={match.params}>
					{match.staticData.breadcrumbTitle}
				</Link>
			</li>
		));

	return <ul>{items}</ul>;
}
```

Add `staticData: { breadcrumbTitle: 'something' }` to any `createFileRoute()` that you want to show up in the breadcrumbs. For example:
```typescript
export const Route = createFileRoute('/post/$year/$month')({
	component: PostsComponent,
	staticData: {
		breadcrumbTitle: 'Posts by month',
	},
});
```

**Reply by damianricobelli** [reply]
I think this is a very smart and simple solution. Thank you

**Reply by CodrinSocol** [reply]
Very elegant, thank you!
